### PR TITLE
Added SimpleBuildReportDeploy script to deploy the build output to target libraries

### PR DESCRIPTION
**Pull request by Anuprakash-Moothedath**
target libraries

Signed-off-by: Anuprakash Moothedath <anuprakash.moothedath@ibm.com>

### CODE REVIEW

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+	System.exit(1)
+}
+
+
+// Read build report data
+def listDirFiles = new File("$tarExtractDirName")
+def buildReportFlag = 0
+
+// Find the BuildReport.json file for deployment 
+listDirFiles.eachFile {
```
Comment (dennis-behm) [inline Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy:111]: Rather than searching for the BuildReport.json in the temp directory, I think we could just assume it to be on the root level for now - this is how the packaging is performed and you make the reference to it.

The logic around a cumulative package of multiple build reports is based on the `buildReportOrder.txt`, but this is fairly complex, while contents of the build reports can overlap. For the first release of this sample script, I would propose, if the buildReportOrder file is found, the script should stop processing and write out to the log that it is not supporting this packaging layout. Further on that, to me this is rather an enhancement of the packaging script to build a cumulative build report as part of its processing, than rebuilding a cumulative process here!

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+   }
+}
+
+// If BuidlReport.json not found in the build output package
+if(!buildReportFlag){
+    println("** Build report data at $tarExtractDirName/BuildReport.json not found")
+}
+
+if (tarExtractDir.exists()) tarExtractDir.deleteDir()
+println("\nDeleted the temporary folder - ${tarExtractDirName}\n")
```
Comment (dennis-behm) [inline Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy:128]: I would rather phrase that as `Cleaning up the temporary folder` . But these are just words

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.properties [hunk 3]
```diff
@@ -0,0 +1,12 @@
+########################################################################
+### The following properties are to configure the untar and deploy
+########################################################################
+
+# This copyModeMap JSON string consists of "deploy type": "the DBB CopyToFS CopyMode" pairs.
+# DBB supported copy modes are documented at
+# https://www.ibm.com/docs/api/v1/content/SS6T76_1.1.0/javadoc/com/ibm/dbb/build/DBBConstants.CopyMode.html
+copyModeMap = {"JCL": "TEXT", "DBRM": "BINARY", "LOAD": "LOAD", "CICSLOAD": "LOAD", "MAPLOAD": "LOAD", "IMSLOAD": "LOAD"}
```
Comment (dennis-behm) [inline Pipeline/SimplePackageDeploy/SimplePackageDeploy.properties:8]: Something which the package actually should contain. Or is this a bad idea?

**Review by dennis-behm — CHANGES_REQUESTED** [review]
Hi Anup, thanks for re-working the SimplePackageDeploy sample script. I left a couple of comments. Maybe we can meet this week to discuss them?

File: Pipeline/SimplePackageDeploy/README.md [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+```
+groovyz SimplePackageDeploy.groovy --workDir /u/ibmuser/workspace/zAppBuild_Hogan/out --tarFileName /u/ibmuser/workspace/zAppBuild_Hogan/out/build.20220920.053231.032/packageWithExtn.tar --hlq IBMUSER.SIT.MORTGAGE --packageWithExtension
+```
+
+<details>
+  <summary>Console log</summary>
+
+SimplePackageDeploy console output
+
+```
```
[the remark this answers is not in the document]
Comment (Anuprakash-Moothedath) [reply Pipeline/SimplePackageDeploy/README.md:84]: Updated zAppBuild_Hogan to MortgageApplication

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy (hunk 1 above)
Comment (Anuprakash-Moothedath) [reply Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy]: Updated the script and readme file for single BuildReport.json as discussed.

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy (hunk 2 above)
Comment (Anuprakash-Moothedath) [reply Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy]: Reworded as advised.

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.properties (hunk 3 above)
Comment (Anuprakash-Moothedath) [reply Pipeline/SimplePackageDeploy/SimplePackageDeploy.properties:8]: As discussed, maintaining this as of now.  Will update once the BuildReport.json from PackageBuildOutputs is updated with copy type.

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+def tarFile = new File("${props.tarFileName}")
+
+if(!tarFile.exists()){
+    println("** Tar file not found at $tarFile")
+    System.exit(1)
+}
+
+
+// Untar the tar file
+println("** Untar file at $tarFile.")
```
Comment (dennis-behm) [inline Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy:94]: I was running your script, and this output in the log confused me.

```
** Untar file at /u/dbehm/test-zapp-app-out/mortgageout/build.20221111.022101.021/build.20221111.022101.021.tar.
Package untar done to /u/dbehm/git/anup/work/DeployFiles_20221111.025913.059
```

I would rather rephrase this to
```
** Untarring file /u/dbehm/test-zapp-app-out/mortgageout/build.20221111.022101.021/build.20221111.022101.021.tar. to /u/dbehm/git/anup/work/DeployFiles_20221111.025913.059
```

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+def buildReportFile = new File("${tarExtractDirName}/BuildReport.json")
+if (buildReportFile.exists()) {	
+	deployFromBuildReport(buildReportFile,tarExtractDirName,targetLibLLQMap,copyModeMap)
+} else {
+	println("** Build report data at $tarExtractDirName/BuildReport.json not found")
+	println("** Deployment stopped")
+}
+
+if (tarExtractDir.exists()) tarExtractDir.deleteDir()
+println("\nCleaning up the temporary folder - ${tarExtractDirName}\n")
```
Comment (dennis-behm) [inline Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy:117]: Please prefix with the `**` to have a better reading of the output log. Also, I don't think we need the newline at the beginning.

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+ *
+ ************************************************************************************/
+
+@Field Properties props = null
+props = parseInput(args)
+
+def startTime = new Date()
+props.startTime = startTime.format("yyyyMMdd.hhmmss.mmm")
+println("** SimplePackageDeploy start at $props.startTime")
+
```
Comment (dennis-behm) [inline Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy:70]: In the PackageBuildOutputs script we also print the parameters/configuration of the script.
```
** Properties at startup:
   verbose -> false
   copyModeMap -> ["COPYBOOK": "TEXT", "COPY": "TEXT", "DBRM": "BINARY", "LOAD": "LOAD"]
   packagingPropertiesFile -> /ZT01/var/dbb/extensions/dbb/Pipeline/PackageBuildOutputs/packageBuildOutputs.properties
   startTime -> 20221111.025209.052
   publish -> false
   workDir -> /u/dbehm/test-zapp-app-out/mortgageout/build.20221111.022101.021/
   buildReportOrder -> [/u/dbehm/test-zapp-app-out/mortgageout/build.20221111.022101.021//BuildReport.json]
   addExtension -> false
```
I would prefer if we have the same for this script as well.

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+	                 println("** Validate if the package was generated with addExtension option. If yes, use option: -e or --packageWithExtension to deploy")
+	                 System.exit(1)
+	             }
+	             copy.setFile(srcFile);
+
+	             copy.setDataset("$targetPDS");
+	             copy.setMember("$member");
+	             copy.copy()
+	         
+	             println("Copied source file - $dataset/$member to Target PDS - $targetPDS")
```
Comment (dennis-behm) [inline Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy:168]: Let's indent the outputs a bit for better reading of the outputs:

Also, I would like to suggest to rephrase the output as:
`Copied file xxx to target library yyy using DBB CopyMode zzz`

```
** Deploying the contents in /u/dbehm/git/anup/work/DeployFiles_20221111.025913.059/BuildReport.json
     Extracted file /u/dbehm/git/anup/work/DeployFiles_20221111.025913.059/DBEHM.DBB.BUILD.DBRM/EPSCMORT is of type DBRM
     Copied source file - DBEHM.DBB.BUILD.DBRM/EPSCMORT to Target PDS - DBEHM.DBB.BUILD.MORTGAGE.DBRM
```

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+	 it.getOutputs().each {
+	     if (it.deployType != null) {
+	         def (dataset, member) = it.dataset.split("\\(|\\)")
+	         def srcFilePath = tarExtractDirName + "/" + dataset + "/" + member		
+	         if (props.packageWithExtension && props.packageWithExtension.toBoolean()) {	
+	         	srcFilePath = srcFilePath + ".${it.deployType}"
+	         }
+	         
+	         def targetPDS = "${props.hlq}" + "." + targetLibLLQMap["${it.deployType}"]
+	         println("\nExtracted file $srcFilePath is of type ${it.deployType}")
```
Comment (dennis-behm) [inline Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy:152]: I don't think that we need this extra output.

The next output line should be sufficient by the reader.

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+	             }
+	             copy.setFile(srcFile);
+
+	             copy.setDataset("$targetPDS");
+	             copy.setMember("$member");
+	             copy.copy()
+	         
+	             println("Copied source file - $dataset/$member to Target PDS - $targetPDS")
+	         } else {
+	             println("ERROR: DEPLOYMENT FAILED")
```
Comment (dennis-behm) [inline Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy:170]: Indentation of text

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+	             println("ERROR: SOURCE FILE NOT DEPLOYED : $dataset/$member")
+	             println("ERROR: DBB COPY MODE NOT DEFINED FOR DEPLOY TYPE : ${it.deployType}\n")
+	             
+	             if (tarExtractDir.exists()) tarExtractDir.deleteDir()
+	             println("\nCleaning up the temporary folder - ${tarExtractDirName}\n")
```
Comment (dennis-behm) [inline Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy:175]: I think, this is a candidate for prefix `**` again.

**Review by dennis-behm — CHANGES_REQUESTED** [review]
@Anuprakash-Moothedath Thanks for integrating the comments we discussed earlier.

The script works nicely. My new comments are targeting the wording and formatting of the outputs to increase readability.

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy (hunk 7 above)
Comment (Anuprakash-Moothedath) [reply Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy:74]: This is done.

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy (hunk 5 above)
Comment (Anuprakash-Moothedath) [reply Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy]: Done

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy (hunk 6 above)
Comment (Anuprakash-Moothedath) [reply Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy]: Done

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy (hunk 8 above)
Comment (Anuprakash-Moothedath) [reply Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy]: Done

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy (hunk 9 above)
Comment (Anuprakash-Moothedath) [reply Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy]: Deleted extra output.

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy (hunk 10 above)
Comment (Anuprakash-Moothedath) [reply Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy]: Done

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy (hunk 11 above)
Comment (Anuprakash-Moothedath) [reply Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy]: Done

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+
+
+// Check if BuildReport.json exists, deploy files if exists and if not stop deploy 
+
+def buildReportFile = new File("${tarExtractDirName}/BuildReport.json")
+if (buildReportFile.exists()) {	
+	deployFromBuildReport(buildReportFile,tarExtractDirName,targetLibLLQMap,copyModeMap)
+} else {
+	println("** Build report data at $tarExtractDirName/BuildReport.json not found")
+	println("** Deployment stopped")
```
Comment (dennis-behm) [inline Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy:116]: In this case, we should also need exit the script, so the pipeline step can be flagged as failed.
`System.exit(1)`

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+	         }
+	         
+	         def targetPDS = "${props.hlq}" + "." + targetLibLLQMap["${it.deployType}"]
+	      
+	         if (copyModeMap["${it.deployType}"] != null) {
+	             copy.setCopyMode(DBBConstants.CopyMode.valueOf(copyModeMap["${it.deployType}"]))
+	         
+	             def srcFile = new File("$srcFilePath")
+	             if(!srcFile.exists()){
+	                 println("** Source file not found at $srcFilePath")
```
Comment (dennis-behm) [inline Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy:161]: Let's use `*!` as an eye-catcher

**Review by dennis-behm — COMMENTED** [review]
@Anuprakash-Moothedath Thanks for the contribution and modifications. This looks very good.

Please see some very tiny change requests.

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy (hunk 12 above)
Comment (Anuprakash-Moothedath) [reply Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy]: Done, Also put eye-catcher *! for the error message.

File: Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy (hunk 13 above)
Comment (Anuprakash-Moothedath) [reply Pipeline/SimplePackageDeploy/SimplePackageDeploy.groovy]: Done

**Review by dennis-behm — APPROVED** [review]
Thanks for the updates!
